### PR TITLE
Turn on TypeScript's strict mode

### LINKED ISSUE #3324: User libraries should be strict by default if written in TypeScript
My opinion is that:

* `.ts` files should be as strict as possible (you can do anything with `as any` trickery still)
* `.js` files should be free to do what ever you want.

Also I think there is no audience who loves working with non-strict TypeScript, it's mainly used by people who are migrating to TypeScript. Catering to "non-strict" TypeScript people is failure in waiting.

(Going in the reverse is more painful, because the libraries will then not type check, if we allow non-strict scripts to prevail.)

Point being that if you use TS you probably love the types and the strictness, if you don't you use JS anyway.

Opening this on suggestion by @kitsonk in #504

### PR DESCRIPTION
**Pull request by ry**

### CODE REVIEW

**Comment by bartlomieju** [comment]
@ry isn't the issue about `strict` option (strict type check), not `alwaysStrict` (`use strict` directive)?

**Comment by rsp** [comment]
@bartlomieju Yes, I believe is that `alwaysStrict` parses in [the JS strict mode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode) (as opposed to  [the JS sloppy mode](https://developer.mozilla.org/en-US/docs/Glossary/Sloppy_mode), not as in strict null checks etc.) and emits `'use strict'` in the output.

The `strict` flag enables `alwaysStrict` **and** `noImplicitAny`, `noImplicitThis`,  `strictBindCallApply`, `strictNullChecks`, `strictFunctionTypes` and `strictPropertyInitialization`.

**Comment by ry** [comment]
Ok fixed.

File: deno_typescript/lib.rs [hunk 1]
```diff
@@ -127,6 +127,7 @@ pub fn compile_bundle(
 
   let config_json = serde_json::json!({
     "compilerOptions": {
+      "strict": true,
```
Comment (kitsonk) [inline deno_typescript/lib.rs]: DOH, I didn't realise that we had eliminated this from here, and that our internal code had drifted so far... 😭 sorry...

**Comment by ry** [comment]
Unfortunately we're going to have to do this in two parts because we reference external code which is not strict:
```
> ./target/debug/deno fmt tests/badly_formatted.js
Compile https://deno.land/std@v0.23.0/prettier/main.ts
error TS7006: Parameter 'opts' implicitly has an 'any' type.

► https://deno.land/std@v0.23.0/prettier/main.ts:334:21

334 async function main(opts): Promise<void> {
                        ~~~~

```

**Comment by rsp** [comment]
> Unfortunately we're going to have to do this in two parts because we reference external code which is not strict

@ry I'd like to help with that. What is the correct way here - patching the std modules or PRs to some upstream repos?

Looking at the copyright notice (Copyright © James Long and contributors) [this particular file](https://deno.land/std@v0.23.0/prettier/main.ts) seems like coming from the prettier project itself but maybe it should me patched directly here.

**Comment by rsp** [comment]
@ry I may be confused here but it seems that contrary to the changelog Deno v0.24.0 doesn't compile TS files with the `--strict` compiler flag. Or am I misunderstanding something here?

I wrote tests for all flags that are turned on by `--strict` (plus some additional ones) here:

https://github.com/rsp/typescript-strict-tests

and the output for Deno v0.24.0 is:
```
Should give all errors for supported flags:
TSC loose:	TSC strict:	Deno:
OK		ERROR		OK		(noImplicitAny)
OK		ERROR		ERROR		(noImplicitThis)
OK		ERROR		ERROR		(alwaysStrict)
OK		ERROR		OK		(strictBindCallApply)
OK		ERROR		OK		(strictNullChecks)
OK		ERROR		OK		(strictFunctionTypes)
OK		ERROR		OK		(strictPropertyInitialization)
OK		ERROR		OK		(noFallthroughCasesInSwitch)
OK		ERROR		OK		(noImplicitReturns)
OK		ERROR		OK		(noUnusedLocals)
OK		ERROR		OK		(noUnusedParameters)
Supported by Deno:    noImplicitThis alwaysStrict
Included by --strict: noImplicitAny noImplicitThis alwaysStrict strictBindCallApply strictNullChecks strictFunctionTypes strictPropertyInitialization
TSC Version 3.7.2 Node v12.13.0
deno: 0.24.0 v8: 8.0.192 typescript: 3.7.2
```

First column means that the test program is OK for TSC without a given flag, second that it gives error with that flag (to make sure that the test works fine) and the third is the result for Deno (OK means that it works like TSC without that flag and ERROR means that it works like with that flag).

Should it behave like this after this PR is merged?

(By the way, maybe those tests will be useful - some minimal programs written to work without strict flags and fails with them: https://github.com/rsp/typescript-strict-tests/tree/master/src)

**Comment by kitsonk** [comment]
What @ry was mentioning is that it needed to be done in two stages... Because the CI process depending on no-strict code, we had to turn strict back on for all the internals, and clean up the problems in `std` so we would have a stable version (0.24) where prettier which we lazy load can be used, so that we can then turn on strict for the runtime.  It is a bit meta...

**Comment by rsp** [comment]
@kitsonk thanks for the clarification.

**Pull request merged**